rust/src/storage/rename.rs

```diff
@@ -0,0 +1,110 @@
+use crate::StorageError;
+use std::ffi::CString;
+
+#[cfg(target_os = "linux")]
+const RENAME_NOREPLACE: libc::c_uint = 1;
+
+#[cfg(target_os = "linux")]
+extern "C" {
+    fn renameat2(
+        olddirfd: libc::c_int,
+        oldpath: *const libc::c_char,
+        newdirfd: libc::c_int,
+        newpath: *const libc::c_char,
+        flags: libc::c_uint,
+    ) -> libc::c_int;
+}
+
+pub fn rename(from: &str, to: &str) -> Result<(), StorageError> {
+    let ret = unsafe {
+        let from = to_c_string(from)?;
+        let to = to_c_string(to)?;
+        platform_specific_rename(from.as_ptr(), to.as_ptr())
+    };
+
+    if ret != 0 {
+        std::fs::remove_file(from).unwrap_or(());
```

Comment:
i feel like it would be less surprise to let caller handle the clean up so semantically it's the same as the rename system call.

Reply:
current design creates tmp file on each try in atomic loop, so if there's concurrent workers it is expected that some of the attemps might now work and hence thos file will remain.

Reply:
But I feel like we can add remove_file only in EEXIST case

Reply:
Right, i think it's fine to leave out temp files, which we will handle later with a gc routine implemented specifically for each storage backend. There are many ways for the writer to leave out dangling temp files, for example, any crash or pause between temp file being created to rename syscall being completed. Moving the temp file remove logic into rename syscall wrapper won't reduce the chance of that happening.

I also think the old behavior you had is more correct :) The file removal should happen on all error cases as long as the writer is not crashed to reduce the dangling temp files we are leaving in the table folder. The writer should still try to be a good citizen when things are in full control :)

Essentially we are implementing a cross platform rename syscall warper for UNIX systems here, so keeping the semantic the same would lead to less surprise to the users. In general, clean up should be handled by the creator of the file because it needs to have file clean code anyway for other failure cases, for example: temp file was created successfully, but file write failed in the middle due to out of disk space.

Reply:
> I also think the old behavior you had is more correct :) The file removal should happen on all error cases 

yeap, but the thing is when we get EEXIST this is expected behavior but rather than an error, since we expected other workers to commit versions concurrently 

> For example: temp file was created successfully, but file write failed in the middle due to out of disk space

This I agree. As mentioned before the clean-up after EEXIST is like using java try-catch for program flow, it's discouraged  but is ok when there's no alternative. 

Anyway, I'm fine with deleting `remove` temp file lines at all, if my statement above does not convince you :)

Reply:
> yeap, but the thing is when we get EEXIST this is expected behavior but rather than an error, since we expected other workers to commit versions concurrently

I agree.

> This I agree. As mentioned before the clean-up after EEXIST is like using java try-catch for program flow, it's discouraged but is ok when there's no alternative.

Yep, catch all error is usually an anti-pattern. I think it just happened that we do need to perform clean up for all the error cases here :( On the other hand, it is a common practice in Linux kernel code to perform this kind of clean up using gotos. Similarly for use of `defer` in golang.

Perhaps it's better to model it as transaction rollback. Creating temp file is one of the step in the transaction. If the transaction failed at the end for whatever error, we need to rollback the full transaction. On transaction commit or abort, all the steps should either all happen or not happen.

Reducing side effect of the function is another way we can justify this behavior. If this function returns, part of the contract we provide should be no dangling temporary file is left on the filesystem.

If leaving those temp files around in error cases provides value to the users, then I think it's worth keeping them. But I don't think it is the case here?

Reply:
> But I don't think it is the case here?

yeah I feel like not.

Overall the transactional approach it is? e.g. to delete tmp file no matter what error

Reply:
yeah, that's what I am leaning towards, unless we come up with a good use for those dangling temp files.